OpenDreamRuntime/Objects/DreamObject.cs

```diff
@@ -106,14 +109,41 @@ public class DreamObject {
             ObjectDefinition = null!;
         }
 
-        public void Delete() {
+
+
+        /// <summary>
+        ///     Enters the current dream object into a global del queue that is guaranteed to run on the DM thread.
+        ///     Use if your deletion handler must be on the DM thread.
+        /// </summary>
+        protected void EnterIntoDelQueue() {
+            DreamManager.DelQueue.Add(this);
+        }
+        /// <summary>
+        ///     Del() the object, cleaning up its variables and refs to minimize size until the .NET GC collects it.
+        /// </summary>
+        /// <param name="possiblyThreaded">If true, Delete() will be defensive and assume it may have been called from another thread by .NET</param>
+        public void Delete(bool possiblyThreaded = false) {
             if (Deleted)
                 return;
 
-            if (TryGetProc("Del", out var delProc))
-                DreamThread.Run(delProc, this, null);
+            if (TryGetProc("Del", out var delProc)) {
+                // SAFETY: See associated comment in Datum.dm. This relies on the invariant that this proc is in a
+                //         thread-safe subset of DM (if such a thing exists) or empty. Currently, it is empty.
+                var datumBaseProc = delProc.OwningType.Path == "/datum";
```

Comment:
```suggestion
                var datumBaseProc = delProc.OwningType == ObjectTree.Datum;
```

Overriding `/datum/Del()` would still lead this to be true. Maybe also check if it's a DMProc with an empty Bytecode?

Reply:
I should probably just add a cached "safe to skip calling" flag to procs, as checking here would be bad for perf.